theano/tensor/nnet/opt.py

```diff
@@ -41,3 +55,299 @@ def local_inplace_sparse_block_outer(node):
+
+    # Conv2d Implementation
+    dx, dy = node.op.subsample
+    if dx not in (1, 2) or dy not in (1, 2):
+        # Not implemented in the gradient of ConvOp
+        return None
+
+    if node.op.imshp is None:
+        op_imshp = (None, None, None, None)
+    else:
+        op_imshp = node.op.imshp
+
+    if node.op.kshp is None:
+        op_kshp = (None, None, None, None)
+    else:
+        op_kshp = node.op.kshp
+
+    if None in op_imshp or None in op_kshp:
+        if (dx, dy) != (1, 1):
+            return None
+
+    mode = 'valid'
+    if not node.op.border_mode == 'full':
+        mode = 'full'
+    filters = kern.dimshuffle((1, 0, 2, 3))
+    filters = filters[:, :, ::-1, ::-1]
+
+    outshp = ConvOp.getOutputShape(op_imshp[2:],
+                                   op_kshp[2:], node.op.subsample,
+                                   node.op.border_mode)
+    fulloutshp = ConvOp.getOutputShape(op_imshp[2:],
+                                       op_kshp[2:], (1, 1),
+                                       node.op.border_mode)
+    nkern = op_imshp[1]
+    imshp = (op_kshp[0], outshp[0], outshp[1])
+    imshp_logical = (op_kshp[0], fulloutshp[0], fulloutshp[1])
+    din = ConvOp(imshp,
+                 op_kshp[2:],
+                 nkern,
+                 op_imshp[0],
+                 1, 1, output_mode=mode,
+                 unroll_batch=None, unroll_kern=None,
+                 unroll_patch=None,
+                 imshp_logical=imshp_logical,
+                 kshp_logical=None,
+                 version=-1,
+                 direction_hint='bprop inputs')
+    din = din(topgrad, filters)
+    din = theano.tensor.patternbroadcast(din, node.outputs[0].broadcastable)
+    return [din]
+
+
+# Register Cpu Optmization
+conv_groupopt = theano.gof.optdb.LocalGroupDB()
+conv_groupopt.__name__ = "conv_opts"
+register_specialize_device(conv_groupopt, 'fast_compile')
+
+# GEMM-based convolution
+# It can be disabled by excluding 'conv_gemm'.
+conv_groupopt.register('local_abstractconv_gemm', local_abstractconv_gemm, 30,
+                       'conv_gemm', 'fast_compile', 'fast_run')
+conv_groupopt.register('local_abstractconv_gradweight_gemm',
+                       local_abstractconv_gradweight_gemm, 30,
+                       'conv_gemm', 'fast_compile', 'fast_run')
+conv_groupopt.register('local_abstractconv_gradinputs_gemm',
+                       local_abstractconv_gradinputs_gemm, 30,
+                       'conv_gemm', 'fast_compile', 'fast_run')
+# Legacy convolution
+conv_groupopt.register('local_conv2d_gpu', local_conv2d_cpu, 40,
+                       'fast_compile', 'fast_run')
+conv_groupopt.register('local_conv2d_gradweight_cpu',
+                       local_conv2d_gradweight_cpu, 40,
+                       'fast_compile', 'fast_run')
+conv_groupopt.register('local_conv2d_gradinputs_cpu',
+                       local_conv2d_gradinputs_cpu, 40,
+                       'fast_compile', 'fast_run')
```

Comment:
I feel like `fast_run` is not needed if we already have `fast_compile`. @nouiz can you confirm?